src/storage/petastorm_storage_engine.py

```diff
@@ -0,0 +1,141 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+
+from src.spark.session import Session
+from src.catalog.models.df_metadata import DataFrameMetadata
+from petastorm.etl.dataset_metadata import materialize_dataset
+from src.storage.abstract_storage_engine import AbstractStorageEngine
+from src.utils.logging_manager import LoggingLevel
+from src.utils.logging_manager import LoggingManager
+from src.configuration.configuration_manager import ConfigurationManager
+
+from petastorm.unischema import dict_to_spark_row
+from petastorm.predicates import in_set, in_lambda
+from petastorm import make_reader
+from typing import Iterator, Dict
+
+
+class PetastormStorageEngine(AbstractStorageEngine):
+
+    def __init__(self):
+        """
+        Maintain a long live spark session and context.
+        """
+        self._spark = Session()
+        self.spark_session = self._spark.get_session()
+        self.spark_context = self._spark.get_context()
+
+    def _spark_url(self, table: DataFrameMetadata) -> str:
+        """
+        Generate a spark/petastorm url given a table
+        """
+        eva_dir = ConfigurationManager().get_value("core", "location")
+        output_url = os.path.join(eva_dir, table.name)
+
+        return output_url
+
+    def create(self, table: DataFrameMetadata):
+        """
+        Create an empty dataframe in petastorm.
+        """
+        empty_rdd = self.spark_context.emptyRDD()
+
+        with materialize_dataset(self.spark_session,
+                                 self._spark_url(table),
+                                 table.schema.petastorm_schema):
+
+            self.spark_session.createDataFrame(empty_rdd,
+                                               table.schema.pyspark_schema) \
+                .coalesce(1) \
+                .write \
+                .mode('overwrite') \
+                .parquet(self._spark_url(table))
+
+
+    def write_row(self, table: DataFrameMetadata, rows: []):
+        """
+        Write rows into the dataframe.
+
+        Arguments:
+            rows List[Dict{key, value}]: We can formally define a dataType Row.
+            keys within the Dict should be consistent with the table.schema.
+        """
+        def row_generator(x):
+            return rows[x]
+
+        with materialize_dataset(self.spark_session,
+                                 self._spark_url(table),
+                                 table.schema.petastorm_schema):
+
+            rows_rdd = self.spark_context.parallelize(range(len(rows)))\
```

Comment:
Directly pass rows as a parameter to parallelize. We can avoid row_generator.

Reply:
I want to do that, but it does not works (or I don't know how to make it work). The reason I think is the spark scheme (generated by petastorm) is binary instead of ndarray, we need go through the petastrom's materialize function to convert ndarray in the row to binary.

Reply:
Directly pass rows as a parameter to parallelize works. Thanks.